core/src/main/java/com/linecorp/armeria/internal/Http2KeepAliveHandler.java

```diff
@@ -0,0 +1,237 @@
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.EventLoop;
+import io.netty.handler.codec.http2.Http2Connection;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.handler.codec.http2.Http2FrameWriter;
+import io.netty.handler.codec.http2.Http2PingFrame;
+import io.netty.handler.timeout.IdleState;
+import io.netty.handler.timeout.IdleStateEvent;
+import io.netty.handler.timeout.IdleStateHandler;
+
+/**
+ * This will send {@link Http2PingFrame} when an {@link IdleStateEvent} is emitted by {@link
+ * IdleStateHandler}. Specifically, it will write a PING frame to remote and then expects an ACK
+ * back within configured {@code pingTimeoutMillis}. If no valid response is received in time, then
+ * channel will be closed.
+ *
+ * <p>This class is <b>not</b> thread-safe and all methods are to be called from single thread such
+ * as {@link EventLoop}.
+ */
+@NotThreadSafe
+public class Http2KeepAliveHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(Http2KeepAliveHandler.class);
+
+    private final boolean sendPingsOnNoActiveStreams;
+    private final long pingTimeoutMillis;
+    private final Http2FrameWriter frameWriter;
+    private final ThreadLocalRandom random = ThreadLocalRandom.current();
+    private final Stopwatch stopwatch = Stopwatch.createUnstarted();
+    private final Http2Connection http2Connection;
+    private final Channel channel;
+    private final ChannelFutureListener pingWriteListener = new PingWriteListener();
+    private final Runnable shutdownRunnable = this::closeChannelAndLog;
+
+    private State state = State.IDLE;
+    private long lastPingPayload;
+    @Nullable
+    private ChannelFuture pingWriteFuture;
+    @Nullable
+    private Future<?> shutdownFuture;
+
+    public Http2KeepAliveHandler(Channel channel, Http2FrameWriter frameWriter,
+                                 Http2Connection http2Connection) {
+        this(channel, frameWriter, http2Connection, Flags.defaultHttp2PingTimeoutMillis(),
+                Flags.defaultUseHttp2PingOnIdleStream());
+    }
+
+    public Http2KeepAliveHandler(Channel channel, Http2FrameWriter frameWriter, Http2Connection http2Connection,
+                                 long pingTimeoutMillis, boolean sendPingsOnNoActiveStreams) {
+        checkArgument(pingTimeoutMillis > 0, pingTimeoutMillis);
+        this.channel = requireNonNull(channel, "channel");
+        this.frameWriter = requireNonNull(frameWriter, "frameWriter");
+        this.pingTimeoutMillis = pingTimeoutMillis;
+        this.http2Connection = requireNonNull(http2Connection, "http2Connection");
+        this.sendPingsOnNoActiveStreams = sendPingsOnNoActiveStreams;
+    }
+
+    public void onChannelIdle(ChannelHandlerContext ctx, IdleStateEvent event) {
+        logger.debug("{} {} event triggered on channel.", channel, event);
```

Comment:
`IdleStateEvent.toString()` returns a string like: `IdleStateEvent(...)`, so the following log message format would be better:

    "{} {} triggered.", channel, event